File: innerlibs1/Encrypt.vb
Imports System.Runtime.CompilerServices
Imports System.Security.Cryptography
Imports System.Text

''' <summary>
''' Modulo de Criptografia
''' </summary>
''' <remarks></remarks>
Public Module Encryption

    ''' <summary>
    ''' Criptografa um Texto em MD5
    ''' </summary>
    ''' <param name="Text">Texto a ser Criptografado</param>
    ''' <returns>Uma String MD5</returns>

    <Extension()>
    Public Function ToMD5String(Text As String) As String
        If (Text.IsNotBlank()) Then
            Dim md5 = System.Security.Cryptography.MD5.Create()
            Dim inputBytes As Byte() = System.Text.Encoding.ASCII.GetBytes(Text)
            Dim hash As Byte() = md5.ComputeHash(inputBytes)
            Dim sb = New StringBuilder()
            For i As Integer = 0 To hash.Length - 1
                sb.Append(hash(i).ToString("X2"))
            Next
            Return sb.ToString()
        End If
        Return Text
    End Function

    ''' <summary>
    ''' Tenta reverter uma string MD5 para seu valor original
    ''' </summary>
    ''' <param name="Text">String MD5</param>
    ''' <returns></returns>
    <Extension()>
    Public Function TryReverseMD5(Text As String) As String
        Return AJAX.Request(Of String)("http://md5.gromweb.com/query/" & Text)
    End Function

    ''' <summary>
    ''' Criptografa uma string
    ''' </summary>
    ''' <param name="Text">Texto descriptografado</param>
    ''' <returns></returns>
    <Extension> Public Function Encrypt(Text As String, Optional Key As String = Nothing) As String
        If Text.IsNotBlank Then

            Dim Results As Byte()
            Dim UTF8 As New System.Text.UTF8Encoding()
            Dim HashProvider As New MD5CryptoServiceProvider()
            Dim TDESKey As Byte() = HashProvider.ComputeHash(UTF8.GetBytes(Key.IfBlank("12345")))
            Dim TDESAlgorithm As New TripleDESCryptoServiceProvider()
            TDESAlgorithm.Key = TDESKey
            TDESAlgorithm.Mode = CipherMode.ECB
            TDESAlgorithm.Padding = PaddingMode.PKCS7
            Dim DataToEncrypt As Byte() = UTF8.GetBytes(Text)
            Try
                Dim Encryptor As ICryptoTransform = TDESAlgorithm.CreateEncryptor()
                Results = Encryptor.TransformFinalBlock(DataToEncrypt, 0, DataToEncrypt.Length)
            Finally
                TDESAlgorithm.Clear()
                HashProvider.Clear()
            End Try
            Return Convert.ToBase64String(Results)
        End If
        Return Text
    End Function

    ''' <summary>
    ''' Descriptografa uma string
    ''' </summary>
    ''' <param name="Text">Texto Criptografado</param>
    ''' <returns></returns>
    <Extension> Public Function Decrypt(Text As String, Optional Key As String = Nothing) As String
        If Text.IsNotBlank Then

            Dim Results As Byte()
            Dim UTF8 As New System.Text.UTF8Encoding()
            Dim HashProvider As New MD5CryptoServiceProvider()
            Dim TDESKey As Byte() = HashProvider.ComputeHash(UTF8.GetBytes(Key.IfBlank("12345")))
            Dim TDESAlgorithm As New TripleDESCryptoServiceProvider()
            TDESAlgorithm.Key = TDESKey
            TDESAlgorithm.Mode = CipherMode.ECB
            TDESAlgorithm.Padding = PaddingMode.PKCS7
            Dim DataToDecrypt As Byte() = Convert.FromBase64String(Text)
            Try
                Dim Decryptor As ICryptoTransform = TDESAlgorithm.CreateDecryptor()
                Results = Decryptor.TransformFinalBlock(DataToDecrypt, 0, DataToDecrypt.Length)
            Finally
                TDESAlgorithm.Clear()
                HashProvider.Clear()
            End Try
            Return UTF8.GetString(Results)
        End If
        Return Text
    End Function

    ''' <summary>
    ''' Criptografa uma string
    ''' </summary>
    ''' <param name="Text">Texto descriptografado</param>
    ''' <returns></returns>
    <Extension()>
    Public Function Encrypt(ByVal text As String, Key As String, IV As String) As String
        If text.IsNotBlank() Then
            Dim aes As AesCryptoServiceProvider = New AesCryptoServiceProvider()
            aes.BlockSize = 128
            aes.KeySize = 256
            aes.IV = Encoding.UTF8.GetBytes(IV)
            aes.Key = Encoding.UTF8.GetBytes(Key)
            aes.Mode = CipherMode.CBC
            aes.Padding = PaddingMode.PKCS7
            Dim src As Byte() = Encoding.UTF8.GetBytes(text)

            Using eencrypt As ICryptoTransform = aes.CreateEncryptor()
                Dim dest As Byte() = eencrypt.TransformFinalBlock(src, 0, src.Length)
                Return Convert.ToBase64String(dest)
            End Using
        End If
        Return text
    End Function

    ''' <summary>
    ''' Descriptografa uma string
    ''' </summary>
    ''' <param name="Text">Texto Criptografado</param>
    ''' <returns></returns>
    <Extension()>
    Public Function Decrypt(ByVal text As String, Key As String, IV As String) As String
        If text.IsNotBlank Then
            Dim aes As AesCryptoServiceProvider = New AesCryptoServiceProvider()
            aes.BlockSize = 128
            aes.KeySize = 256
            aes.IV = Encoding.UTF8.GetBytes(IV)
            aes.Key = Encoding.UTF8.GetBytes(Key)
            aes.Mode = CipherMode.CBC
            aes.Padding = PaddingMode.PKCS7
            Dim src As Byte()
            Try
                src = System.Convert.FromBase64String(text.FixBase64())
            Catch
                src = System.Convert.FromBase64String(text)

            End Try

            Using ddecrypt As ICryptoTransform = aes.CreateDecryptor()
                Try
                    Dim dest As Byte() = ddecrypt.TransformFinalBlock(src, 0, src.Length)
                    Return Encoding.UTF8.GetString(dest)
                Catch ex As Exception
                End Try
            End Using
        End If
        Return text
    End Function

End Module

File: InnerLibs/Encrypt.cs
using System;
using System.Globalization;
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace InnerLibs
{
    /// <summary>
    /// Modulo de Criptografia
    /// </summary>
    /// <remarks></remarks>
    public static partial class Util
    {
        /// <summary>
        /// Criptografa um Texto em MD5
        /// </summary>
        /// <param name="Text">Texto a ser Criptografado</param>
        /// <returns>Uma String MD5</returns>

        #region Public Methods

        /// <summary>
        /// Descriptografa uma string
        /// </summary>
        /// <param name="Text">Texto Criptografado</param>
        /// <returns></returns>
        public static string Decrypt(this string Text, string Key = null)
        {
            if (Text.IsNotBlank())
            {
                byte[] Results = default;
                var UTF8 = new UTF8Encoding();
                var HashProvider = new MD5CryptoServiceProvider();
                var TDESKey = HashProvider.ComputeHash(UTF8.GetBytes(Key.IfBlank("1234567890")));
                var TDESAlgorithm = new TripleDESCryptoServiceProvider
                {
                    Key = TDESKey,
                    Mode = CipherMode.ECB,
                    Padding = PaddingMode.PKCS7
                };
                var DataToDecrypt = Convert.FromBase64String(Text);
                try
                {
                    var Decryptor = TDESAlgorithm.CreateDecryptor();
                    Results = Decryptor.TransformFinalBlock(DataToDecrypt, 0, DataToDecrypt.Length);
                }
                finally
                {
                    TDESAlgorithm.Clear();
                    HashProvider.Clear();
                }
                TDESAlgorithm.Dispose();
                HashProvider.Dispose();
                return UTF8.GetString(Results);
            }

            return Text;
        }

        /// <summary>
        /// Descriptografa uma string
        /// </summary>
        /// <param name="Text">Texto Criptografado</param>
        /// <returns></returns>
        public static string Decrypt(this string text, string Key, string IV)
        {
            if (text.IsNotBlank())
            {
                var aes = new AesCryptoServiceProvider
                {
                    BlockSize = 128,
                    KeySize = 256,
                    IV = new UTF8Encoding(false).GetBytes(IV),
                    Key = new UTF8Encoding(false).GetBytes(Key),
                    Mode = CipherMode.CBC,
                    Padding = PaddingMode.PKCS7
                };
                byte[] src;
                try
                {
                    src = Convert.FromBase64String(text.FixBase64());
                }
                catch
                {
                    src = Convert.FromBase64String(text);
                }

                using (var ddecrypt = aes.CreateDecryptor())
                {
                    try
                    {
                        var dest = ddecrypt.TransformFinalBlock(src, 0, src.Length);
                        return new UTF8Encoding(false).GetString(dest);
                    }
                    catch
                    {
                    }
                }
            }

            return text;
        }

        public static FileInfo DecryptRSA(this FileInfo File, string Key) => File?.ToBytes().DecryptRSA(Key).WriteToFile(File.FullName);

        /// <summary>
        /// Descriptografa um array de bytes encriptada em RSA
        /// </summary>
        /// <param name="bytes"></param>
        /// <param name="Key"></param>
        /// <returns></returns>
        public static byte[] DecryptRSA(this byte[] bytes, string Key)
        {
            var cspp = new CspParameters() { KeyContainerName = Key };
            var rsa = new RSACryptoServiceProvider(cspp) { PersistKeyInCsp = true };
            bytes = rsa.Decrypt(bytes, true);
            rsa.Dispose();
            return bytes;
        }

        /// <summary>
        /// Criptografa uma string
        /// </summary>
        /// <param name="Text">Texto descriptografado</param>
        /// <returns></returns>
        public static string Encrypt(this string Text, string Key = null)
        {
            if (Text.IsNotBlank())
            {
                byte[] Results = default;
                var UTF8 = new UTF8Encoding();
                var HashProvider = new MD5CryptoServiceProvider();
                var TDESKey = HashProvider.ComputeHash(UTF8.GetBytes(Key.IfBlank("1234567890")));
                var TDESAlgorithm = new TripleDESCryptoServiceProvider
                {
                    Key = TDESKey,
                    Mode = CipherMode.ECB,
                    Padding = PaddingMode.PKCS7
                };
                var DataToEncrypt = UTF8.GetBytes(Text);
                try
                {
                    var Encryptor = TDESAlgorithm.CreateEncryptor();
                    Results = Encryptor.TransformFinalBlock(DataToEncrypt, 0, DataToEncrypt.Length);
                }
                finally
                {
                    TDESAlgorithm.Clear();
                    HashProvider.Clear();
                }
                TDESAlgorithm.Dispose();
                HashProvider.Dispose();
                return Convert.ToBase64String(Results);
            }

            return Text;
        }

        /// <summary>
        /// Criptografa uma string
        /// </summary>
        /// <param name="Text">Texto descriptografado</param>
        /// <returns></returns>
        public static string Encrypt(this string text, string Key, string IV)
        {
            if (text.IsNotBlank())
            {
                var aes = new AesCryptoServiceProvider
                {
                    BlockSize = 128,
                    KeySize = 256,
                    IV = new UTF8Encoding(false).GetBytes(IV),
                    Key = new UTF8Encoding(false).GetBytes(Key),
                    Mode = CipherMode.CBC,
                    Padding = PaddingMode.PKCS7
                };
                var src = new UTF8Encoding(false).GetBytes(text);
                using (var eencrypt = aes.CreateEncryptor())
                {
                    var dest = eencrypt.TransformFinalBlock(src, 0, src.Length);
                    return Convert.ToBase64String(dest);
                }
            }

            return text;
        }

        /// <summary>
        /// Criptografa um string em RSA
        /// </summary>
        /// <param name="Text"></param>
        /// <param name="Key"></param>
        /// <returns></returns>
        public static string EncryptRSA(this string Text, string Key)
        {
            var cspp = new CspParameters() { KeyContainerName = Key };
            var rsa = new RSACryptoServiceProvider(cspp) { PersistKeyInCsp = true };
            var bytes = rsa.Encrypt(new UTF8Encoding(false).GetBytes(Text), true);
            rsa.Dispose();
            return BitConverter.ToString(bytes);
        }

        /// <summary>
        /// Criptografa um array de bytes em RSA
        /// </summary>
        /// <param name="bytes"></param>
        /// <param name="Key"></param>
        /// <returns></returns>
        public static byte[] EncryptRSA(this byte[] bytes, string Key)
        {
            var cspp = new CspParameters() { KeyContainerName = Key };
            using (var rsa = new RSACryptoServiceProvider(cspp) { PersistKeyInCsp = true })
            {
                return rsa.Encrypt(bytes, true);
            }
        }

        public static FileInfo EncryptRSA(this FileInfo File, string Key) => File?.ToBytes().EncryptRSA(Key).WriteToFile(File.FullName);

        public static string ToMD5String(this string Text)
        {
            if (Text.IsNotBlank())
            {
                var md5 = MD5.Create();
                var inputBytes = Encoding.ASCII.GetBytes(Text);
                var hash = md5.ComputeHash(inputBytes);
                var sb = new StringBuilder();
                for (int i = 0, loopTo = hash.Length - 1; i <= loopTo; i++)
                {
                    sb.Append(hash[i].ToString("X2", CultureInfo.InvariantCulture));
                }

                return sb.ToString();
            }

            return Text;
        }

        #endregion Public Methods
    }
}